File: src/main/java/ch/zxseitz/j3de/math/Vector3.java
package ch.zxseitz.j3de.math;

public class Vector3 {
    public static final Vector3 ZERO = new Vector3(0f, 0f, 0f);
    public static final Vector3 ONE = new Vector3(1f, 1f, 1f);
    public static final Vector3 XY = new Vector3(1f, 1f, 0f);
    public static final Vector3 XZ = new Vector3(1f, 0f, 1f);
    public static final Vector3 YZ = new Vector3(0f, 1f, 1f);
    public static final Vector3 X = new Vector3(1f, 0f, 0f);
    public static final Vector3 Y = new Vector3(0f, 1f, 0f);
    public static final Vector3 Z = new Vector3(0f, 0f, 1f);

    public final float x, y, z;

    public Vector3(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3(Vector3 vec) {
        this.x = vec.x;
        this.y = vec.y;
        this.z = vec.z;
    }

    public Vector3(Vector2 vec) {
        this.x = vec.x;
        this.y = vec.y;
        this.z = 0;
    }

    public Vector3 add(Vector3 vec) {
        return new Vector3(
                x + vec.x,
                y + vec.y,
                z + vec.z
        );
    }

    public Vector3 subtract(Vector3 vec) {
        return new Vector3(
                x - vec.x,
                y - vec.y,
                z - vec.z
        );
    }

    public Vector3 multiply(Vector3 vec) {
        return new Vector3(
                x * vec.x,
                y * vec.y,
                z * vec.z
        );
    }

    public Vector3 divide(Vector3 vec) {
        return new Vector3(
                x / vec.x,
                y / vec.y,
                z / vec.z
        );
    }

    public Vector3 scale(float s) {
        return new Vector3(
                x * s,
                y * s,
                z * s
        );
    }

    public float norm() {
        return (float) Math.sqrt(normSquared());
    }

    public float normSquared() {
        return x * x + y * y + z * z;
    }

    public float dot(Vector3 vec) {
        return x * vec.x + y * vec.y + z * vec.z;
    }

    public Vector3 cross(Vector3 vec) {
        return new Vector3(
                y * vec.z - z * vec.y,
                z * vec.x - x * vec.z,
                x * vec.y - y * vec.x
        );
    }

    public Vector4 toHomogenous() {
        return new Vector4(x, y, z, 1f);
    }

    @Override
    public boolean equals(Object obj) {
//        if (obj instanceof Vector3 vec) {
        if (obj instanceof Vector3) {
            var vec = (Vector3) obj;
            return MathUtils.isFloatEquals(x, vec.x) && MathUtils.isFloatEquals(y, vec.y)
                    && MathUtils.isFloatEquals(z, vec.z);
        }
        return false;
    }

    @Override
    public String toString() {
        return String.format("(%.3f, %.3f, %.3f)", x, y, z);
    }
}

File: src/main/java/ch/zxseitz/j3de/math/Vector3.kt
package ch.zxseitz.j3de.math

import kotlin.math.sqrt

data class Vector3(val x: Float = 0f, val y: Float = 0f, val z: Float = 0f) {
    companion object {
        val ZERO = Vector3()
        val ONE = Vector3(1f, 1f, 1f)
        val XY = Vector3(1f, 1f, 0f)
        val XZ = Vector3(1f, 0f, 1f)
        val YZ = Vector3(0f, 1f, 1f)
        val X = Vector3(1f, 0f, 0f)
        val Y = Vector3(0f, 1f, 0f)
        val Z = Vector3(0f, 0f, 1f)
    }

    constructor(v: Vector2, z: Float = 0f) : this(v.x, v.y, z)

    operator fun plus(v: Vector3): Vector3 {
        return Vector3(x + v.x, y + v.y, z + v.z)
    }

    operator fun minus(v: Vector3): Vector3 {
        return Vector3(x - v.x, y - v.y, z - v.z)
    }

    operator fun times(s: Float): Vector3 {
        return Vector3(x * s, y * s, z * s)
    }

    fun normSquared(): Float {
        return x * x + y * y + z * z
    }

    fun norm(): Float {
        return sqrt(normSquared())
    }

    fun normalize(): Vector3 {
        return this * (1f / norm())
    }

    fun toHomogen(): Vector4 {
        return Vector4(this, 1f)
    }

    override fun toString(): String {
        return "(%.3f, %.3f, %.3f)".format(x, y, z)
    }
}

operator fun Vector3.unaryMinus() = Vector3(-x, -y, -z)

infix fun Vector3._times(v: Vector3): Vector3 {
    return Vector3(x * v.x, y * v.y, z * v.z)
}

infix fun Vector3.dot(v: Vector3): Float {
    return x * v.x + y * v.y + z * v.z
}

infix fun Vector3.cross(v: Vector3): Vector3 {
    return Vector3(
        y * v.z - z * v.y,
        z * v.x - x * v.z,
        x * v.y - y * v.x
    )
}
